google-ads/src/test/java/com/google/ads/googleads/lib/utils/FieldMasksTest.java

```diff
@@ -1,24 +1,102 @@
+import static org.junit.Assert.assertThrows;
 
-import com.google.ads.googleads.v5.resources.Campaign;
+import com.google.ads.googleads.v6.enums.CampaignStatusEnum.CampaignStatus;
+import com.google.ads.googleads.v6.resources.Campaign;
 import com.google.common.collect.ImmutableList;
+import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.JUnit4;
 
 @RunWith(JUnit4.class)
 public class FieldMasksTest {
+  private Campaign campaign;
+
+  @Before
+  public void setUp() {
+    Campaign.Builder builder =
+        Campaign.newBuilder()
+            .setName("New campaign")
+            .setStatus(CampaignStatus.PAUSED);
+    builder.getManualCpcBuilder().setEnhancedCpcEnabled(false);
+    campaign = builder.build();
+  }
 
   // Regression test for https://github.com/googleads/google-ads-java/issues/344.
   @Test
   public void optionalField_withDefaultValue_detectsChange() {
     Campaign.Builder builder = Campaign.newBuilder();
     builder.getManualCpcBuilder().setEnhancedCpcEnabled(false);
-    Campaign campaign = builder.build();
+    campaign = builder.build();
     assertEquals(
         ImmutableList.of("manual_cpc.enhanced_cpc_enabled"),
         FieldMasks.allSetFieldsOf(campaign).getPathsList());
   }
-}
+
+  @Test
+  public void getFieldValue_gets_basic_field() {
+    assertEquals(FieldMasks.getFieldValue("name", campaign), campaign.getName());
+  }
+
+  @Test
+  public void getFieldValue_gets_nested_field() {
+    assertEquals(
+        FieldMasks.getFieldValue("manual_cpc.enhanced_cpc_enabled", campaign),
+        campaign.getManualCpc().getEnhancedCpcEnabled());
+  }
+
+  @Test
+  public void getFieldValue_gets_top_nested_field() {
+    assertEquals(FieldMasks.getFieldValue("manual_cpc", campaign), campaign.getManualCpc());
+  }
+
+  @Test
+  public void getFieldValue_gets_enum_field() {
+    assertEquals(
+        FieldMasks.getFieldValue("status", campaign), campaign.getStatus().getValueDescriptor());
+  }
+
+  @Test
+  public void getFieldValue_no_match_throws_exception() {
+    IllegalArgumentException exception =
+        assertThrows(
+            IllegalArgumentException.class,
+            () -> {
+              FieldMasks.getFieldValue("foo.bar", campaign);
+            });
+    assertEquals(
+        exception.getMessage(),
+        "Cannot retrieve field value. A matching field was not found after navigating foo.bar up to"
+            + " foo.");
+  }
+
+  @Test
+  public void getFieldValue_repeated_throws_exception() {
+    IllegalStateException exception =
+        assertThrows(
+            IllegalStateException.class,
+            () -> {
+              FieldMasks.getFieldValue("url_custom_parameters.key", campaign);
+            });
+    assertEquals(
+        exception.getMessage(),
+        "Cannot retrieve field value. A repeated field was encountered after navigating"
+            + " url_custom_parameters.key up to url_custom_parameters.");
+  }
+
+  @Test
```

Comment:
Can you add a test case for attempting to access a subfield of a Message when the containing message is null.

Reply:
I couldn't figure out a way to test this, but I'm open to ideas. There are very few cases where a value can actually be null. For example, [campaign.optimization_score](https://developers.google.com/google-ads/api/fields/v6/campaign#campaign.optimization_score) can be null, but this field is output only. Without digging through the entire API, it seems output only fields are the only field that could possibly have null values. With the `Campaign` created in the test, the optimization score defaults to 0.0. Moreover, you can't mock a resource on account of the fact that resources are `final`. That's the reason for the missing test coverage here, but I'm open to ideas.

Reply:
That's a good point, in proto3 even message fields are never null.